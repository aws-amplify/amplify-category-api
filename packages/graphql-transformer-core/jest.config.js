const baseConfig = require('../../jest.config.base.js'); // eslint-disable-line @typescript-eslint/no-var-requires

module.exports = {
  ...baseConfig,
  coverageThreshold: {
    global: {
      branches: 65,
      functions: 33,
      lines: 47,
    },
  },
};
