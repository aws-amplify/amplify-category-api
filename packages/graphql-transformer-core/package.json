{
  "name": "graphql-transformer-core",
  "version": "8.2.4",
  "description": "A framework to transform from GraphQL SDL to AWS cloudFormation.",
  "repository": {
    "type": "git",
    "url": "https://github.com/aws-amplify/amplify-category-api.git",
    "directory": "packages/graphql-transformer-core"
  },
  "author": "Amazon Web Services",
  "license": "Apache-2.0",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "keywords": [
    "graphql",
    "cloudformation",
    "aws"
  ],
  "scripts": {
    "test": "jest",
    "build": "rimraf ./lib ./tsconfig.tsbuildinfo && tsc",
    "watch": "tsc -w",
    "clean": "rimraf ./lib",
    "extract-api": "ts-node ../../scripts/extract-api.ts"
  },
  "dependencies": {
    "cloudform-types": "^4.2.0",
    "deep-diff": "^1.0.2",
    "fs-extra": "^8.1.0",
    "glob": "^10.3.0",
    "graphql": "^15.5.0",
    "graphql-transformer-common": "4.27.1",
    "lodash": "^4.17.21"
  },
  "devDependencies": {
    "@aws-amplify/amplify-cli-core": "^4.2.10",
    "@types/fs-extra": "^8.0.1",
    "@types/node": "^18.17.0",
    "mock-fs": "5.2.0",
    "@types/mock-fs": "4.13.4"
  },
  "jest": {
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    },
    "testURL": "http://localhost",
    "testRegex": "(src/__tests__/.*|(\\.|/)(test|spec))\\.(jsx?|tsx?)$",
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js",
      "jsx",
      "json",
      "node"
    ],
    "collectCoverage": true,
    "coverageProvider": "v8",
    "coverageThreshold": {
      "global": {
        "branches": 67,
        "functions": 33,
        "lines": 48
      }
    },
    "coverageReporters": [
      "clover",
      "text"
    ]
  }
}
