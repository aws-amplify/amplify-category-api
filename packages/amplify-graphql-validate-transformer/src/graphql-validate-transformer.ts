import { ValidateDirective } from '@aws-amplify/graphql-directives';
import { DirectiveWrapper, MappingTemplate, TransformerPluginBase } from '@aws-amplify/graphql-transformer-core';
import {
  TransformerContextProvider,
  TransformerSchemaVisitStepContextProvider,
  TransformerPluginProvider,
} from '@aws-amplify/graphql-transformer-interfaces';
import { DirectiveNode, FieldDefinitionNode, InterfaceTypeDefinitionNode, ObjectTypeDefinitionNode } from 'graphql';
import { ValidateArguments, ValidateDirectiveConfiguration } from './types';
import { validate } from './validators';
import { makeValidationSnippet } from './vtl-generator';

export class ValidateTransformer extends TransformerPluginBase implements TransformerPluginProvider {
  private directiveMap = new Map<string, ValidateDirectiveConfiguration[]>();

  constructor() {
    super('amplify-graphql-validate-transformer', ValidateDirective.definition);
  }

  /**
   * Processes `@validate` directives and validates their usage and inputs.
   *
   * @param parent - The parent object or interface type definition node containing the field
   * @param definition - The field definition node being processed
   * @param directive - The `@validate` directive node applied to the field
   * @param _ - The transformer schema visit step context provider (unused)
   */
  field = (
    parent: ObjectTypeDefinitionNode | InterfaceTypeDefinitionNode,
    definition: FieldDefinitionNode,
    directive: DirectiveNode,
    _: TransformerSchemaVisitStepContextProvider,
  ): void => {
    const directiveWrapped = new DirectiveWrapper(directive);
    const config = this.getValidateDirectiveConfiguration(directiveWrapped, parent as ObjectTypeDefinitionNode, definition);

    validate(parent as ObjectTypeDefinitionNode, definition, directive, config);

    const parentName = parent.name.value;
    if (!this.directiveMap.has(parentName)) {
      this.directiveMap.set(parentName, []);
    }
    this.directiveMap.get(parentName)!.push(config);
  };

  /**
   * Generates resolvers for validation directives.
   *
   * @param ctx - The transformer context provider
   */
  generateResolvers = (ctx: TransformerContextProvider): void => {
    const mutationTypeName = ctx.output.getMutationTypeName();
    if (!mutationTypeName) {
      return;
    }

    for (const typeName of this.directiveMap.keys()) {
      for (const config of this.directiveMap.get(typeName)!) {
        const fieldName = config.fieldNode.name.value;
        const validationType = config.type;
        const validationValue = config.value;
        const errorMessage = config.errorMessage || `Validation failed for ${fieldName}`; // TODO: default error message
        const validationSnippet = makeValidationSnippet(fieldName, validationType, validationValue, errorMessage);

        // Add validation to create mutation
        const createFieldName = `create${typeName}`;
        const createResolver = ctx.resolvers.getResolver(mutationTypeName, createFieldName);
        if (createResolver) {
          createResolver.addVtlFunctionToSlot(
            'validate',
            MappingTemplate.s3MappingTemplateFromString(
              validationSnippet,
              `${mutationTypeName}.${createFieldName}.{slotName}.{slotIndex}.req.vtl`,
            ),
          );
        }

        // Add validation to update mutation
        const updateFieldName = `update${typeName}`;
        const updateResolver = ctx.resolvers.getResolver(mutationTypeName, updateFieldName);
        if (updateResolver) {
          updateResolver.addVtlFunctionToSlot(
            'validate',
            MappingTemplate.s3MappingTemplateFromString(
              validationSnippet,
              `${mutationTypeName}.${updateFieldName}.{slotName}.{slotIndex}.req.vtl`,
            ),
          );
        }
      }
    }
  };

  /**
   * Extracts and formats the validation directive configuration from a `@validate` directive.
   *
   * @param directive - The wrapped directive containing validation arguments
   * @param parentNode - The object type definition node that contains the field
   * @param fieldNode - The field definition node that the directive is applied to
   * @returns A configuration object containing the validation rules and metadata
   * @private
   */
  private getValidateDirectiveConfiguration(
    directive: DirectiveWrapper,
    parentNode: ObjectTypeDefinitionNode,
    fieldNode: FieldDefinitionNode,
  ): ValidateDirectiveConfiguration {
    const defaultArgs: ValidateArguments = {
      type: '',
      value: '',
      errorMessage: '',
    };
    const args = directive.getArguments<ValidateArguments>(defaultArgs);

    return {
      parentNode,
      fieldNode,
      type: args.type,
      value: args.value,
      errorMessage: args.errorMessage,
    };
  }
}
