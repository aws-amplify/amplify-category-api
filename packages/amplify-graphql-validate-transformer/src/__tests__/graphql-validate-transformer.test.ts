import { ModelTransformer } from '@aws-amplify/graphql-model-transformer';
import { validateModelSchema } from '@aws-amplify/graphql-transformer-core';
import { parse } from 'graphql';
import { testTransform } from '@aws-amplify/graphql-transformer-test-utils';
import { ValidateTransformer } from '../graphql-validate-transformer';

type ValidationCheck = {
  field: string;
  message: string;
};

describe('ValidateTransformer', () => {
  const transformSchema = (schema: string): any => {
    const out = testTransform({
      schema,
      transformers: [new ModelTransformer(), new ValidateTransformer()],
    });
    expect(out).toBeDefined();
    validateModelSchema(parse(out.schema));
    return out;
  };

  const verifyValidations = (out: any, type: string, validations: ValidationCheck[]): void => {
    validations.forEach((validation, index) => {
      const createMutation = out.resolvers[`Mutation.create${type}.validate.${index + 1}.req.vtl`];
      expect(createMutation).toBeDefined();
      expect(createMutation).toContain(validation.message);

      const updateMutation = out.resolvers[`Mutation.update${type}.validate.${index + 1}.req.vtl`];
      expect(updateMutation).toBeDefined();
      expect(updateMutation).toContain(validation.message);
    });
  };

  it('should successfully transform schema with numeric validations', () => {
    const schema = `
      type Product @model {
        id: ID!
        price: Float! @validate(type: gt, value: "0", errorMessage: "Price must be positive")
          @validate(type: lt, value: "1000000", errorMessage: "Price must be less than 1,000,000")
        quantity: Int! @validate(type: gte, value: "0", errorMessage: "Quantity cannot be negative")
          @validate(type: lte, value: "100", errorMessage: "Quantity cannot exceed 100")
      }`;

    const out = transformSchema(schema);

    verifyValidations(out, 'Product', [
      { field: 'price', message: 'Price must be positive' },
      { field: 'price', message: 'Price must be less than 1,000,000' },
      { field: 'quantity', message: 'Quantity cannot be negative' },
      { field: 'quantity', message: 'Quantity cannot exceed 100' },
    ]);
  });

  it('should successfully transform schema with string validations', () => {
    const schema = `
      type User @model {
        id: ID!
        email: String! @validate(type: matches, value: "^[A-Za-z0-9+_.-]+@(.+)$", errorMessage: "Invalid email format")
        username: String! @validate(type: minLength, value: "3", errorMessage: "Username too short")
          @validate(type: maxLength, value: "20", errorMessage: "Username too long")
        url: String @validate(type: startsWith, value: "https://", errorMessage: "URL must start with https://")
          @validate(type: endsWith, value: ".com", errorMessage: "URL must end with .com")
      }`;

    const out = transformSchema(schema);

    verifyValidations(out, 'User', [
      { field: 'email', message: 'Invalid email format' },
      { field: 'username', message: 'Username too short' },
      { field: 'username', message: 'Username too long' },
      { field: 'url', message: 'URL must start with https://' },
      { field: 'url', message: 'URL must end with .com' },
    ]);
  });

  it('should allow multiple validations on the same field', () => {
    const schema = `
      type Post @model {
        id: ID!
        title: String! @validate(type: minLength, value: "5", errorMessage: "Title too short")
          @validate(type: maxLength, value: "100", errorMessage: "Title too long")
          @validate(type: startsWith, value: "prefix", errorMessage: "Title must start with prefix")
          @validate(type: endsWith, value: "suffix", errorMessage: "Title must end with suffix")
          @validate(type: matches, value: "^[A-Za-z0-9 ]+$", errorMessage: "Title can only contain letters, numbers and spaces")
      }`;

    const out = transformSchema(schema);

    verifyValidations(out, 'Post', [
      { field: 'title', message: 'Title too short' },
      { field: 'title', message: 'Title too long' },
      { field: 'title', message: 'Title must start with prefix' },
      { field: 'title', message: 'Title must end with suffix' },
      { field: 'title', message: 'Title can only contain letters, numbers and spaces' },
    ]);
  });

  it('should successfully transform schema with validations on multiple fields', () => {
    const schema = `
      type Comment @model {
        id: ID!
        content: String! @validate(type: minLength, value: "1", errorMessage: "Content cannot be empty")
          @validate(type: maxLength, value: "1000", errorMessage: "Content too long")
        author: String! @validate(type: minLength, value: "3", errorMessage: "Author too short")
          @validate(type: maxLength, value: "20", errorMessage: "Author too long")
        rating: Int! @validate(type: gte, value: "0", errorMessage: "Rating cannot be negative")
          @validate(type: lte, value: "5", errorMessage: "Rating cannot exceed 5")
      }`;

    const out = transformSchema(schema);

    verifyValidations(out, 'Comment', [
      { field: 'content', message: 'Content cannot be empty' },
      { field: 'content', message: 'Content too long' },
      { field: 'author', message: 'Author too short' },
      { field: 'author', message: 'Author too long' },
      { field: 'rating', message: 'Rating cannot be negative' },
      { field: 'rating', message: 'Rating cannot exceed 5' },
    ]);
  });
});
