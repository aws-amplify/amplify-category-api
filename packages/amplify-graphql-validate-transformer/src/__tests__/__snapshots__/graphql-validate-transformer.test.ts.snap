// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValidateTransformer should generate correct validation resolvers for multiple fields: create comment author maxLength validation 1`] = `
"## [Start] Validating \\"author\\" with type \\"maxLength\\" and value \\"20\\". **
#if( !$util.isNull($ctx.args.input.author) )
      #set($validationPassed = $$ctx.args.input.author.length() <= 20)
      #if(!$validationPassed)
        $util.error(\\"Author too long\\")
      #end
    #end
## [End] Validating \\"author\\" with type \\"maxLength\\" and value \\"20\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for multiple fields: create comment author minLength validation 1`] = `
"## [Start] Validating \\"author\\" with type \\"minLength\\" and value \\"3\\". **
#if( !$util.isNull($ctx.args.input.author) )
      #set($validationPassed = $$ctx.args.input.author.length() >= 3)
      #if(!$validationPassed)
        $util.error(\\"Author too short\\")
      #end
    #end
## [End] Validating \\"author\\" with type \\"minLength\\" and value \\"3\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for multiple fields: create comment content maxLength validation 1`] = `
"## [Start] Validating \\"content\\" with type \\"maxLength\\" and value \\"1000\\". **
#if( !$util.isNull($ctx.args.input.content) )
      #set($validationPassed = $$ctx.args.input.content.length() <= 1000)
      #if(!$validationPassed)
        $util.error(\\"Content too long\\")
      #end
    #end
## [End] Validating \\"content\\" with type \\"maxLength\\" and value \\"1000\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for multiple fields: create comment content minLength validation 1`] = `
"## [Start] Validating \\"content\\" with type \\"minLength\\" and value \\"1\\". **
#if( !$util.isNull($ctx.args.input.content) )
      #set($validationPassed = $$ctx.args.input.content.length() >= 1)
      #if(!$validationPassed)
        $util.error(\\"Content cannot be empty\\")
      #end
    #end
## [End] Validating \\"content\\" with type \\"minLength\\" and value \\"1\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for multiple fields: create comment rating gte validation 1`] = `
"## [Start] Validating \\"rating\\" with type \\"gte\\" and value \\"0\\". **
#if( !$util.isNull($ctx.args.input.rating) )
      #set($validationPassed = $$ctx.args.input.rating >= 0)
      #if(!$validationPassed)
        $util.error(\\"Rating cannot be negative\\")
      #end
    #end
## [End] Validating \\"rating\\" with type \\"gte\\" and value \\"0\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for multiple fields: create comment rating lte validation 1`] = `
"## [Start] Validating \\"rating\\" with type \\"lte\\" and value \\"5\\". **
#if( !$util.isNull($ctx.args.input.rating) )
      #set($validationPassed = $$ctx.args.input.rating <= 5)
      #if(!$validationPassed)
        $util.error(\\"Rating cannot exceed 5\\")
      #end
    #end
## [End] Validating \\"rating\\" with type \\"lte\\" and value \\"5\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for multiple fields: update comment author maxLength validation 1`] = `
"## [Start] Validating \\"author\\" with type \\"maxLength\\" and value \\"20\\". **
#if( !$util.isNull($ctx.args.input.author) )
      #set($validationPassed = $$ctx.args.input.author.length() <= 20)
      #if(!$validationPassed)
        $util.error(\\"Author too long\\")
      #end
    #end
## [End] Validating \\"author\\" with type \\"maxLength\\" and value \\"20\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for multiple fields: update comment author minLength validation 1`] = `
"## [Start] Validating \\"author\\" with type \\"minLength\\" and value \\"3\\". **
#if( !$util.isNull($ctx.args.input.author) )
      #set($validationPassed = $$ctx.args.input.author.length() >= 3)
      #if(!$validationPassed)
        $util.error(\\"Author too short\\")
      #end
    #end
## [End] Validating \\"author\\" with type \\"minLength\\" and value \\"3\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for multiple fields: update comment content maxLength validation 1`] = `
"## [Start] Validating \\"content\\" with type \\"maxLength\\" and value \\"1000\\". **
#if( !$util.isNull($ctx.args.input.content) )
      #set($validationPassed = $$ctx.args.input.content.length() <= 1000)
      #if(!$validationPassed)
        $util.error(\\"Content too long\\")
      #end
    #end
## [End] Validating \\"content\\" with type \\"maxLength\\" and value \\"1000\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for multiple fields: update comment content minLength validation 1`] = `
"## [Start] Validating \\"content\\" with type \\"minLength\\" and value \\"1\\". **
#if( !$util.isNull($ctx.args.input.content) )
      #set($validationPassed = $$ctx.args.input.content.length() >= 1)
      #if(!$validationPassed)
        $util.error(\\"Content cannot be empty\\")
      #end
    #end
## [End] Validating \\"content\\" with type \\"minLength\\" and value \\"1\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for multiple fields: update comment rating gte validation 1`] = `
"## [Start] Validating \\"rating\\" with type \\"gte\\" and value \\"0\\". **
#if( !$util.isNull($ctx.args.input.rating) )
      #set($validationPassed = $$ctx.args.input.rating >= 0)
      #if(!$validationPassed)
        $util.error(\\"Rating cannot be negative\\")
      #end
    #end
## [End] Validating \\"rating\\" with type \\"gte\\" and value \\"0\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for multiple fields: update comment rating lte validation 1`] = `
"## [Start] Validating \\"rating\\" with type \\"lte\\" and value \\"5\\". **
#if( !$util.isNull($ctx.args.input.rating) )
      #set($validationPassed = $$ctx.args.input.rating <= 5)
      #if(!$validationPassed)
        $util.error(\\"Rating cannot exceed 5\\")
      #end
    #end
## [End] Validating \\"rating\\" with type \\"lte\\" and value \\"5\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for multiple validations on the same field: create post title endsWith validation 1`] = `
"## [Start] Validating \\"title\\" with type \\"endsWith\\" and value \\"suffix\\". **
#if( !$util.isNull($ctx.args.input.title) )
      #set($validationPassed = $$ctx.args.input.title.endsWith(\\"suffix\\"))
      #if(!$validationPassed)
        $util.error(\\"Title must end with suffix\\")
      #end
    #end
## [End] Validating \\"title\\" with type \\"endsWith\\" and value \\"suffix\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for multiple validations on the same field: create post title matches validation 1`] = `
"## [Start] Validating \\"title\\" with type \\"matches\\" and value \\"^[A-Za-z0-9 ]+$\\". **
#if( !$util.isNull($ctx.args.input.title) )
      #set($validationPassed = $util.matches($$ctx.args.input.title, \\"^[A-Za-z0-9 ]+$\\"))
      #if(!$validationPassed)
        $util.error(\\"Title can only contain letters, numbers and spaces\\")
      #end
    #end
## [End] Validating \\"title\\" with type \\"matches\\" and value \\"^[A-Za-z0-9 ]+$\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for multiple validations on the same field: create post title maxLength validation 1`] = `
"## [Start] Validating \\"title\\" with type \\"maxLength\\" and value \\"100\\". **
#if( !$util.isNull($ctx.args.input.title) )
      #set($validationPassed = $$ctx.args.input.title.length() <= 100)
      #if(!$validationPassed)
        $util.error(\\"Title too long\\")
      #end
    #end
## [End] Validating \\"title\\" with type \\"maxLength\\" and value \\"100\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for multiple validations on the same field: create post title minLength validation 1`] = `
"## [Start] Validating \\"title\\" with type \\"minLength\\" and value \\"5\\". **
#if( !$util.isNull($ctx.args.input.title) )
      #set($validationPassed = $$ctx.args.input.title.length() >= 5)
      #if(!$validationPassed)
        $util.error(\\"Title too short\\")
      #end
    #end
## [End] Validating \\"title\\" with type \\"minLength\\" and value \\"5\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for multiple validations on the same field: create post title startsWith validation 1`] = `
"## [Start] Validating \\"title\\" with type \\"startsWith\\" and value \\"prefix\\". **
#if( !$util.isNull($ctx.args.input.title) )
      #set($validationPassed = $$ctx.args.input.title.startsWith(\\"prefix\\"))
      #if(!$validationPassed)
        $util.error(\\"Title must start with prefix\\")
      #end
    #end
## [End] Validating \\"title\\" with type \\"startsWith\\" and value \\"prefix\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for multiple validations on the same field: update post title endsWith validation 1`] = `
"## [Start] Validating \\"title\\" with type \\"endsWith\\" and value \\"suffix\\". **
#if( !$util.isNull($ctx.args.input.title) )
      #set($validationPassed = $$ctx.args.input.title.endsWith(\\"suffix\\"))
      #if(!$validationPassed)
        $util.error(\\"Title must end with suffix\\")
      #end
    #end
## [End] Validating \\"title\\" with type \\"endsWith\\" and value \\"suffix\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for multiple validations on the same field: update post title matches validation 1`] = `
"## [Start] Validating \\"title\\" with type \\"matches\\" and value \\"^[A-Za-z0-9 ]+$\\". **
#if( !$util.isNull($ctx.args.input.title) )
      #set($validationPassed = $util.matches($$ctx.args.input.title, \\"^[A-Za-z0-9 ]+$\\"))
      #if(!$validationPassed)
        $util.error(\\"Title can only contain letters, numbers and spaces\\")
      #end
    #end
## [End] Validating \\"title\\" with type \\"matches\\" and value \\"^[A-Za-z0-9 ]+$\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for multiple validations on the same field: update post title maxLength validation 1`] = `
"## [Start] Validating \\"title\\" with type \\"maxLength\\" and value \\"100\\". **
#if( !$util.isNull($ctx.args.input.title) )
      #set($validationPassed = $$ctx.args.input.title.length() <= 100)
      #if(!$validationPassed)
        $util.error(\\"Title too long\\")
      #end
    #end
## [End] Validating \\"title\\" with type \\"maxLength\\" and value \\"100\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for multiple validations on the same field: update post title minLength validation 1`] = `
"## [Start] Validating \\"title\\" with type \\"minLength\\" and value \\"5\\". **
#if( !$util.isNull($ctx.args.input.title) )
      #set($validationPassed = $$ctx.args.input.title.length() >= 5)
      #if(!$validationPassed)
        $util.error(\\"Title too short\\")
      #end
    #end
## [End] Validating \\"title\\" with type \\"minLength\\" and value \\"5\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for multiple validations on the same field: update post title startsWith validation 1`] = `
"## [Start] Validating \\"title\\" with type \\"startsWith\\" and value \\"prefix\\". **
#if( !$util.isNull($ctx.args.input.title) )
      #set($validationPassed = $$ctx.args.input.title.startsWith(\\"prefix\\"))
      #if(!$validationPassed)
        $util.error(\\"Title must start with prefix\\")
      #end
    #end
## [End] Validating \\"title\\" with type \\"startsWith\\" and value \\"prefix\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for numeric validations: create product price gt validation 1`] = `
"## [Start] Validating \\"price\\" with type \\"gt\\" and value \\"0\\". **
#if( !$util.isNull($ctx.args.input.price) )
      #set($validationPassed = $$ctx.args.input.price > 0)
      #if(!$validationPassed)
        $util.error(\\"Price must be positive\\")
      #end
    #end
## [End] Validating \\"price\\" with type \\"gt\\" and value \\"0\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for numeric validations: create product price lt validation 1`] = `
"## [Start] Validating \\"price\\" with type \\"lt\\" and value \\"1000000\\". **
#if( !$util.isNull($ctx.args.input.price) )
      #set($validationPassed = $$ctx.args.input.price < 1000000)
      #if(!$validationPassed)
        $util.error(\\"Price must be less than 1,000,000\\")
      #end
    #end
## [End] Validating \\"price\\" with type \\"lt\\" and value \\"1000000\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for numeric validations: create product quantity gte validation 1`] = `
"## [Start] Validating \\"quantity\\" with type \\"gte\\" and value \\"0\\". **
#if( !$util.isNull($ctx.args.input.quantity) )
      #set($validationPassed = $$ctx.args.input.quantity >= 0)
      #if(!$validationPassed)
        $util.error(\\"Quantity cannot be negative\\")
      #end
    #end
## [End] Validating \\"quantity\\" with type \\"gte\\" and value \\"0\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for numeric validations: create product quantity lte validation 1`] = `
"## [Start] Validating \\"quantity\\" with type \\"lte\\" and value \\"100\\". **
#if( !$util.isNull($ctx.args.input.quantity) )
      #set($validationPassed = $$ctx.args.input.quantity <= 100)
      #if(!$validationPassed)
        $util.error(\\"Quantity cannot exceed 100\\")
      #end
    #end
## [End] Validating \\"quantity\\" with type \\"lte\\" and value \\"100\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for numeric validations: update product price gt validation 1`] = `
"## [Start] Validating \\"price\\" with type \\"gt\\" and value \\"0\\". **
#if( !$util.isNull($ctx.args.input.price) )
      #set($validationPassed = $$ctx.args.input.price > 0)
      #if(!$validationPassed)
        $util.error(\\"Price must be positive\\")
      #end
    #end
## [End] Validating \\"price\\" with type \\"gt\\" and value \\"0\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for numeric validations: update product price lt validation 1`] = `
"## [Start] Validating \\"price\\" with type \\"lt\\" and value \\"1000000\\". **
#if( !$util.isNull($ctx.args.input.price) )
      #set($validationPassed = $$ctx.args.input.price < 1000000)
      #if(!$validationPassed)
        $util.error(\\"Price must be less than 1,000,000\\")
      #end
    #end
## [End] Validating \\"price\\" with type \\"lt\\" and value \\"1000000\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for numeric validations: update product quantity gte validation 1`] = `
"## [Start] Validating \\"quantity\\" with type \\"gte\\" and value \\"0\\". **
#if( !$util.isNull($ctx.args.input.quantity) )
      #set($validationPassed = $$ctx.args.input.quantity >= 0)
      #if(!$validationPassed)
        $util.error(\\"Quantity cannot be negative\\")
      #end
    #end
## [End] Validating \\"quantity\\" with type \\"gte\\" and value \\"0\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for numeric validations: update product quantity lte validation 1`] = `
"## [Start] Validating \\"quantity\\" with type \\"lte\\" and value \\"100\\". **
#if( !$util.isNull($ctx.args.input.quantity) )
      #set($validationPassed = $$ctx.args.input.quantity <= 100)
      #if(!$validationPassed)
        $util.error(\\"Quantity cannot exceed 100\\")
      #end
    #end
## [End] Validating \\"quantity\\" with type \\"lte\\" and value \\"100\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for string validations: create user email matches validation 1`] = `
"## [Start] Validating \\"email\\" with type \\"matches\\" and value \\"^[A-Za-z0-9+_.-]+@(.+)$\\". **
#if( !$util.isNull($ctx.args.input.email) )
      #set($validationPassed = $util.matches($$ctx.args.input.email, \\"^[A-Za-z0-9+_.-]+@(.+)$\\"))
      #if(!$validationPassed)
        $util.error(\\"Invalid email format\\")
      #end
    #end
## [End] Validating \\"email\\" with type \\"matches\\" and value \\"^[A-Za-z0-9+_.-]+@(.+)$\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for string validations: create user url endsWith validation 1`] = `
"## [Start] Validating \\"url\\" with type \\"endsWith\\" and value \\".com\\". **
#if( !$util.isNull($ctx.args.input.url) )
      #set($validationPassed = $$ctx.args.input.url.endsWith(\\".com\\"))
      #if(!$validationPassed)
        $util.error(\\"URL must end with .com\\")
      #end
    #end
## [End] Validating \\"url\\" with type \\"endsWith\\" and value \\".com\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for string validations: create user url startsWith validation 1`] = `
"## [Start] Validating \\"url\\" with type \\"startsWith\\" and value \\"https://\\". **
#if( !$util.isNull($ctx.args.input.url) )
      #set($validationPassed = $$ctx.args.input.url.startsWith(\\"https://\\"))
      #if(!$validationPassed)
        $util.error(\\"URL must start with https://\\")
      #end
    #end
## [End] Validating \\"url\\" with type \\"startsWith\\" and value \\"https://\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for string validations: create user username maxLength validation 1`] = `
"## [Start] Validating \\"username\\" with type \\"maxLength\\" and value \\"20\\". **
#if( !$util.isNull($ctx.args.input.username) )
      #set($validationPassed = $$ctx.args.input.username.length() <= 20)
      #if(!$validationPassed)
        $util.error(\\"Username too long\\")
      #end
    #end
## [End] Validating \\"username\\" with type \\"maxLength\\" and value \\"20\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for string validations: create user username minLength validation 1`] = `
"## [Start] Validating \\"username\\" with type \\"minLength\\" and value \\"3\\". **
#if( !$util.isNull($ctx.args.input.username) )
      #set($validationPassed = $$ctx.args.input.username.length() >= 3)
      #if(!$validationPassed)
        $util.error(\\"Username too short\\")
      #end
    #end
## [End] Validating \\"username\\" with type \\"minLength\\" and value \\"3\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for string validations: update user email matches validation 1`] = `
"## [Start] Validating \\"email\\" with type \\"matches\\" and value \\"^[A-Za-z0-9+_.-]+@(.+)$\\". **
#if( !$util.isNull($ctx.args.input.email) )
      #set($validationPassed = $util.matches($$ctx.args.input.email, \\"^[A-Za-z0-9+_.-]+@(.+)$\\"))
      #if(!$validationPassed)
        $util.error(\\"Invalid email format\\")
      #end
    #end
## [End] Validating \\"email\\" with type \\"matches\\" and value \\"^[A-Za-z0-9+_.-]+@(.+)$\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for string validations: update user url endsWith validation 1`] = `
"## [Start] Validating \\"url\\" with type \\"endsWith\\" and value \\".com\\". **
#if( !$util.isNull($ctx.args.input.url) )
      #set($validationPassed = $$ctx.args.input.url.endsWith(\\".com\\"))
      #if(!$validationPassed)
        $util.error(\\"URL must end with .com\\")
      #end
    #end
## [End] Validating \\"url\\" with type \\"endsWith\\" and value \\".com\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for string validations: update user url startsWith validation 1`] = `
"## [Start] Validating \\"url\\" with type \\"startsWith\\" and value \\"https://\\". **
#if( !$util.isNull($ctx.args.input.url) )
      #set($validationPassed = $$ctx.args.input.url.startsWith(\\"https://\\"))
      #if(!$validationPassed)
        $util.error(\\"URL must start with https://\\")
      #end
    #end
## [End] Validating \\"url\\" with type \\"startsWith\\" and value \\"https://\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for string validations: update user username maxLength validation 1`] = `
"## [Start] Validating \\"username\\" with type \\"maxLength\\" and value \\"20\\". **
#if( !$util.isNull($ctx.args.input.username) )
      #set($validationPassed = $$ctx.args.input.username.length() <= 20)
      #if(!$validationPassed)
        $util.error(\\"Username too long\\")
      #end
    #end
## [End] Validating \\"username\\" with type \\"maxLength\\" and value \\"20\\". **"
`;

exports[`ValidateTransformer should generate correct validation resolvers for string validations: update user username minLength validation 1`] = `
"## [Start] Validating \\"username\\" with type \\"minLength\\" and value \\"3\\". **
#if( !$util.isNull($ctx.args.input.username) )
      #set($validationPassed = $$ctx.args.input.username.length() >= 3)
      #if(!$validationPassed)
        $util.error(\\"Username too short\\")
      #end
    #end
## [End] Validating \\"username\\" with type \\"minLength\\" and value \\"3\\". **"
`;
